### PR TITLE
Performance: use event delegation

### PR DESCRIPTION
**Pull request by Indigo744**
Ok, this one is quite big.

Until now, we attached to each area/plot/link node the following event listeners:
 - mouseover (for hover attributes and tooltips)
 - mousemove (for tooltip position)
 - mouseout (for hover attributes and tooltips)
 - click (for href)

This can become quite a lot for big maps with lots of elements.

The solution is actually quite simple: use event delegation!
Basically, instead of attaching all event listener per node, we attach only one event listener to the container, and filter the child element with delegation.
More information: http://api.jquery.com/on/ and https://learn.jquery.com/events/event-delegation/

So the modifications are:
 - remove any event attachment for nodes (`$().on()..`)
 - implements each event listeners on container in `initDelegatedMapEvents()` and dispatch relevant action

~~For now, I haven't touched the way the custom `eventHandlers` works. They are still attached to each node according to the user option. What do you think?~~

Here is a test using [one of the example with lots of plots and tooltips](https://rawgit.com/neveldo/jQuery-Mapael/2.1.0/examples/basic/legend_plotted_cities.html)
**Before:**
Huge number of listener events
1113 events listeners, 21 MB JS heap
**After:**
Small number of listener events
4 events listeners, 15 MB JS heap

This may help for #304

I've checked all example without finding any regression. But I'll let you review this ;-)

### CODE REVIEW

**Comment by Indigo744** [comment]
Ah, I may need to update the tests

**Comment by neveldo** [comment]
Hi @Indigo744 ,

Wow, indeed its a big one. Thanks a lot for it, I was having in my mind this issue regarding the huge number of events added by Mapael since a long time, but never took the time to dive into it !

I will take the necessary time to look at your PR and I will keep you informated within the next few days !

**Comment by Indigo744** [comment]
Added delegation for custom events!
I've namespaced them like 'click.mapael.custom' to be able to easily unbind/rebind them.

**Comment by neveldo** [comment]
Wow thanks ! I keep you informed ASAP for the review !

File: js/jquery.mapael.js [hunk 1]
```diff
@@ -361,6 +364,83 @@
...
+            var self = this;
+
+            /* Attach mouseover event delegation
+             * Note: we filter the event with a timeout to reduce the firing when the mouse moves quickly
+             */
+            var mapMouseOverTimeoutID;
+            self.$map.on("mouseover." + pluginName, "[data-id]", function () {
+                var elem = this;
+                clearTimeout(mapMouseOverTimeoutID);
+                mapMouseOverTimeoutID = setTimeout(function(){
```
Comment (neveldo) [inline js/jquery.mapael.js]: missing space between ) and {

File: js/jquery.mapael.js [hunk 2]
```diff
@@ -361,6 +364,83 @@
...
+            });
+
+            /* Attach mousemove event delegation
+             * Note: timeout filtering is small to update the Tooltip position fast
+             */
+            var mapMouseMoveTimeoutID;
+            self.$map.on("mousemove." + pluginName, "[data-id]", function (event) {
+                var elem = this;
+                clearTimeout(mapMouseMoveTimeoutID);
+                mapMouseMoveTimeoutID = setTimeout(function(){
```
Comment (neveldo) [inline js/jquery.mapael.js]: missing space between ) and {

File: js/jquery.mapael.js [hunk 3]
```diff
@@ -1880,87 +1871,129 @@
...
-            else mapElem.attr(mapElem.attrsHover);
-            // Set textElem
-            if (textElem) {
-                if (textElem.attrsHover.animDuration > 0) textElem.animate(textElem.attrsHover, textElem.attrsHover.animDuration);
-                else textElem.attr(textElem.attrsHover);
+            if (elem === undefined) return;
+
+            /* Handle tooltip position update */
+            if (elem.tooltip !== undefined) {
+                console.log(elem);
```
Comment (neveldo) [inline js/jquery.mapael.js]: console.log() should be removed

File: js/jquery.mapael.js (hunk 1 above)
Comment (Indigo744) [reply js/jquery.mapael.js]: Done

File: js/jquery.mapael.js (hunk 3 above)
Comment (Indigo744) [reply js/jquery.mapael.js]: I don't see it actually... Did you check the latest rev?

File: js/jquery.mapael.js [hunk 4]
```diff
@@ -438,6 +438,29 @@
...
+            var mapClickTimeoutID;
+            self.$map.on("click." + pluginName, "[data-id]", function () {
+                var elem = this;
+                clearTimeout(mapClickTimeoutID);
+                mapClickTimeoutID = setTimeout(function(){
+                    var $elem = $(elem);
+                    var id = $elem.attr('data-id');
+                    var type = $elem.attr('data-type');
+
+                    if (type === 'area' || type === 'area-text') {
```
[the remark this answers is not in the document]
Comment (Indigo744) [reply js/jquery.mapael.js]: Really good idea! I just committed the change.

File: js/jquery.mapael.js [hunk 5]
```diff
@@ -1954,7 +1952,7 @@
                     if (elem.tooltip.overflow.right === true) {
                         tooltipPosition.left = mouseX - self.$map.offset().left + 10;
                     }
-                    if (selem.tooltip.overflow.bottom === true) {
+                    if (elem.tooltip.overflow.bottom === true) {
```
[the remark this answers is not in the document]
Comment (Indigo744) [reply js/jquery.mapael.js]: Ahah yes indeed!

**Comment by Indigo744** [comment]
Since we are already talking about v3 (plugin related), I'm wondering if we should postpone this evolution to the next major release?

I am afraid it may break on some people who has weirdly override some event handlers on each element... 

What do you think?

**Comment by neveldo** [comment]
Hi @Indigo744 ,

Thanks for the fixes and for having separated the work in small commits, it helped a lot for the review ! (I agree for the squash when we will merge it with the master branch).

Have you some weirdly override cases that will not work anymore when this PR will be merged ? 
Don't you think that all regular use cases should still work fine after this PR will be merged ?

I think that if some user extend some of the internal functions of Mapael, they should do this with full knowledge of the facts.Otherwise, each minor change of any internal function interface should lead to a new major release of mapael.

**Comment by Indigo744** [comment]
Nah I don't have any weird override cases ^^ but I was trying to imagine what could be the implication.

I think you are right, and the benefits can be important for users with lots of plots.

As a side note, I was wondering if we should put the filtering Timeout values as a variable so user can override them if needed.

**Comment by neveldo** [comment]
I think it could be indeed a good idea to make the timeout delay an internal variable to allow overriding. However, I think we shouldn't expose it as an regular option as the great majority of users will not need to override it.

**Pull request merged**